src/Common/Core/Impl/Services/ServiceManager.cs

```diff
@@ -0,0 +1,149 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Common.Core.Diagnostics;
+
+namespace Microsoft.Common.Core.Services {
+    public class ServiceManager : IServiceManager {
+        class ServiceFactory {
+            public IServiceFactory Factory { get; set; }
+            public object[] Parameters { get; set; }
+        }
+
+        private readonly Dictionary<Type, object> _services = new Dictionary<Type, object>();
+        private readonly Dictionary<Type, ServiceFactory> _deferredServices = new Dictionary<Type, ServiceFactory>();
```

Comment:
I originally thought that factories from `_deferredServices` create new object for each call, but it looks like I got it wrong and they are also one per type. In this case, if instead of `IServiceFactory` instance and an array of parameters you provide a simple delegate `Func<T> factory`, you can store it in the dictionary as `new Lazy<T>(factory), and then access it`s value in `GetService`. Another benefit of it is that dictionary don't have to be locked while factory is executed for the first time - only callers that need service of this particular type will wait.